src/network/client.rs

```diff
@@ -230,13 +349,82 @@ impl NetworkClient {
     }
 
     fn handle_heartbeat(&self, peer_id: PeerId, heartbeat: Heartbeat) {
-        if !heartbeat.version_matches(&SUPPORTED_VERSIONS) {
-            metrics::IGNORED_PINGS.inc();
+        tracing::debug!("Ping from {peer_id} {:?}", heartbeat.assignment_id);
+        let local_assignments = self.assignments.lock();
+        let Some(assignment) = local_assignments.get(&heartbeat.assignment_id) else {
+            tracing::debug!("Assignment {:?} not found", heartbeat.assignment_id);
+            let latest_assignmaent_id = local_assignments
+                .last_key_value()
+                .map(|(assignment_id, _)| assignment_id.clone())
+                .unwrap_or_default();
+            if heartbeat.assignment_id > latest_assignmaent_id {
+                self.heartbeat_buffer.lock().push_back((peer_id, heartbeat));
```

Comment:
This is a hot event loop, so locks should be avoided if possible. Can we do `try_lock` and add the heartbeat to the buffer if locking fails (since we have the buffer anyway)?